service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -940,6 +940,43 @@ protected void onCommandConnectionEstablished(final HonoConnection commandConnec
             final String deviceId,
             final Device authenticatedDevice,
             final SpanContext context) {
+        return getRegistrationAssertion(tenantId, deviceId, authenticatedDevice, context, true);
+    }
+
+    /**
+     * Gets an assertion for a device's registration status.
+     * <p>
+     * The returned JSON object may include <em>default</em>
+     * values for properties to set on messages published by the device under
+     * property {@link RegistrationConstants#FIELD_PAYLOAD_DEFAULTS}.
+     * <p>
+     * Note that this method will also update the last gateway associated with the given device if
+     * <em>updateLastGateway</em> is {@code true}.
+     *
+     * @param tenantId The tenant that the device belongs to.
+     * @param deviceId The device to get the assertion for.
+     * @param authenticatedDevice The device that has authenticated to this protocol adapter.
+     *            <p>
+     *            If not {@code null} then the authenticated device is compared to the given tenant and device ID. If
+     *            they differ in the device identifier, then the authenticated device is considered to be a gateway
+     *            acting on behalf of the device.
+     * @param context The currently active OpenTracing span that is used to
+     *                trace the retrieval of the assertion.
+     * @param updateLastGateway If {@code true}, the last gateway associated with the given device will be updated.
+     *                          Setting this value to {@code false} and thereby skipping the update makes sense for
+     *                          protocol adapters where gateways subscribe to commands for a specific device (not
+     *                          for all devices connected to the gateway). This is usually the case for adapters
```

Comment:
This is usually the case for adapters that use a stateless protocol like HTTP for communication with devices. In such a case, the adapter will usually delay the response to a device's request for the time indicated by the *ttd* request property in order to prevent buffering of commands in the adapter.

Reply:
> the adapter will usually delay the response to a device's request for the time indicated by the ttd request property

Is that really the case? To me it looks like a received command at the HTTP adapter will directly be sent to the device as part of the HTTP response - without further delay.
And preventing further commands to be received and buffered is done by not giving back any further credit to the messaging network.

In any case, what the "This is usually.." sentence should do IMHO, is to make clear why a adapter would want to only support commands for *specific devices*, instead of *gateways*. In your version this doesn't become so clear to me.

Reply:
> To me it looks like a received command at the HTTP adapter will directly be sent to the device as part of the HTTP response - without further delay.

The important thing here is that the adapter needs to associate a command for a device with a gateway's HTTP request on behalf of that specific device. The way we have implemented the command forwarding in the HTTP response assumes that a command contained in the HTTP response is targeted at the device that the request was referring to. Thus, there is no way for the adapter to send a command to a device outside of the scope of an HTTP request from (or on behalf of) that device.

Alternatively, the adapter could be modified to support *long polling* of gateways, i.e. a gateway could consecutively POST empty notifications to `/event`, thus indicating that the gateway itself is ready to receive commands. The adapter could then indeed just open a single command receiver for the gateway and forward commands for devices in the HTTP responses to the gateway. However, this would require us to include the target device ID in the response (a header) and also this would probably not work too well because the gateway might send each of its *long polling* requests to a different adapter instance. In any case, the adapter would then also need to update a device's via information on every message ...

So, I do not think that it is about the amount of time to wait but more about the fact that the HTTP adapter (currently) is not able to include commands for arbitrary devices in HTTP responses for a gateway. BTW we have the same problem with the MQTT adapter, I guess, because there we also (currently) cannot include the target device ID in a command message being published to a command receiver topic (which therefore includes the device either explicitly or implicitly).

Reply:
@calohmn is this still the way to go in context of #1552 ?

Reply:
@sophokles73 No, this PR becomes obsolete with #1552.